### DISCUSSION TITLE
Improve OIDC key management

### DISCUSSION THREAD

**Discussion by muhlemmer** [zitadel/zitadel#7464] [Ideas]
Private keys are used in ZITADEL to sign JSON Web Tokens after an OIDC Authentication flow is completed successfully. These are usual ID Tokens and access tokens if the application is configured to use JWT. Public keys are used to verify these tokens inside and outside of ZITADEL. The [`jwks_uri`](https://zitadel.com/docs/apis/openidoauth/endpoints#jwks_uri`) can be called by applications to obtain active public keys in the form of a JSON Web Key Set (JWKS).

## Current management

Currently keys are managed based on expiry. Expiry of keys is not required by any standard. Our initial idea behind this mechanism is that we force frequent rotation of keys and therefore teach app developers to deal with that situation by anticipating key rotation as part of their business logic and not something that has to be dealt with incidentally. 

### Private keys

Initially a ZITADEL instance has no active key pair.  When the OIDC handler needs to sign a token, it tries to look for a valid private key. As it can't find one a new key pair is generated, stored and used to sign the token. The private key is then used until we pass the time of expiry. 

The database only returns non-expired private keys. So when the private key is expired, ZITADEL won't find a key and will start the cycle from the beginning, generating a new key pair.

By default and on ZITADEL cloud private keys expire after 6 hours.

https://github.com/zitadel/zitadel/blob/f4c72cbe144e5cbbfcc17c638143bf2b3f5656f7/cmd/defaults.yaml#L502

### Public keys

Public keys are generated along with private keys and stored. In ZITADEL we expire public keys in order to keep the result set on the `jwks_ur` in check. There is no need to return keys that are not used by any tokens anymore. 

By default and on ZITADEL cloud public keys expire after 30 hours.

https://github.com/zitadel/zitadel/blob/f4c72cbe144e5cbbfcc17c638143bf2b3f5656f7/cmd/defaults.yaml#L503

Token lifetime is 12 hours by default, but can be modified on instance level.

https://github.com/zitadel/zitadel/blob/f4c72cbe144e5cbbfcc17c638143bf2b3f5656f7/cmd/defaults.yaml#L711-L712

An active instance running on a ZITADEL system with above configuration can have as many as 5 active public keys while 1 private key is currently used for signing. However, an instance that has been inactive for 30 hours will not have any active public keys will have an empty result set when calling the `jwks_uri`.

## Community feedback

The current management solution has let to a number of discussions, misunderstandings and people being plain unhappy about the situation. Even internally at ZITADEL we sometimes had to find creative solutions to work around expired public keys.

### Bug reports

Users of low-traffic instance sometimes assume something is broken when `jwks_uri` returns no keys. Arguments sometimes include that we do not comply to some standard (disputed):
- https://github.com/zitadel/zitadel/discussions/7198
- https://github.com/zitadel/zitadel/issues/7206
- https://github.com/zitadel/zitadel/issues/6949#issuecomment-1857759664
- https://discord.com/channels/927474939156643850/1203978361643204628
- https://discord.com/channels/927474939156643850/1203980489585729547
- https://discord.com/channels/927474939156643850/1178641311323136020

### Breaking applications

Some applications middleware seem to pre-fetch keys. Those apps might encounter tokens signed with a key id for which they do not have private key yet. Although we recommend that applications call the `jwks_uri` when that happens, there are situations where developers don't have that level of control over the middleware that might be enforced by a framework.

- https://github.com/zitadel/zitadel/discussions/7198#discussioncomment-8133299
- https://discord.com/channels/927474939156643850/1193964899844505650

### Seamless key rotation

There was a previous discussion that proposed seamless rotation that got some positive feedback and up-votes. Unfortunately it got closed by the OP.

- https://github.com/zitadel/zitadel/discussions/5690

### Userinfo / introspection

When the access token lifetime on instance level is more than then public key lifetime from the zitadel runtime config (30 hrs in `defaults.yaml`), userinfo and introspection will fail when JWTs are used.

- https://discord.com/channels/927474939156643850/1232959992953442386. This user configured 36hrs token lifetime on zitadel cloud, where 30hrs is the public key lifetime.

## Proposal

**Create a resource API for keys** where a client with sufficient permissions can:

- Create a new keypair
- Set a private key as the current signing key
- Delete keypairs for forced rotation
- Get/List keypair metadata (creation date, active etc)

**Initial keypair** created and stored on instance creation

**Console UI** to be extended using the API, so that admins can easily manage keys from there.

**Keys never expire**. They must be deleted manually.

### Pros

- All public keys of not-deleted key pairs are **always** returned from the `jwks_uri`. No more empty responses.
- Allows admins to create a keypair ahead of time and "promote" it to the signing key at any later time. This gives the admin the freedom to make the interval long enough for the new public key to propagate to his apps. 
- Makes the state of keypairs visible to the admin (currently it's just a black box)
- Allows to be extended to have a choice of signing algorithm. This is a thought for the future and outside the scope of this proposal.

### Cons

- Users that rely on auto-rotation will need to start managing keys themselves
- When the new behavior is enforced globally, it will be a breaking change wrt key lifetime behavior.

#### Consolidation

- Cloud platforms or *nix based servers have provisions for running time based commands. Like `cron` or [Google Cloud Scheduler](https://cloud.google.com/scheduler/docs/overview). There are also [Teraform examples](https://medium.com/@manish.aganani/terraform-schedule-a-cron-job-to-call-a-cloud-function-20c96f8dd295) on the internet.
- We now have [Feature Flags](https://github.com/zitadel/zitadel/pull/7356) so we can put the new behavior behind an opt-in flag.

## Finishing thoughts

The purpose of this discussion is to poll the community if there are people against this proposal, if improvements can be made to it and if there is sufficient demand, keeping the [original upvotes](https://github.com/zitadel/zitadel/discussions/5690) in mind, of course. 

If you like the proposal as-is, please give it a thumbs-up or upvote. Feel free to comment about anything you wish to discuss further.

**Comment by holotrack** [comment]
I will just add here that we get in to troubles because of it when using Zitadel and Istio together. Examples of related issues showing that Istio wont handle it on their site (there are more of them):

https://github.com/istio/istio/issues/29436
https://github.com/istio/istio/issues/37710
https://github.com/istio/istio/issues/49044

You can setup with Istio only refresh interval. Istio itself wont refresh key when not presented. We set it up for 1min and we still getting information from users that problem appear from time to time.

Forcing developers to use refreshed keys maybe was good idea but what with users which are using some other solutions (eg Istio in that case) which not implemented it and doesn't look they will? 

I think there should be setting to optional out from this refresh feature when needed to let peoples choice


EDIT:
![image](https://github.com/zitadel/zitadel/assets/100580892/401fda92-df16-4a4a-ac30-12ec4564f786)

Maybe i dont know about something but why not both. You already implemented rotation so leave it but give possibility to turn it off or at least set up some different value than 6 hours eg once per century

**Reply by muhlemmer** [reply]
We could still provide the auto-rotation on zitadel cloud with a cron-job. But I wouldn't keep it in the zitadel business logic in the future.

We have "feature flags" for zitadel now, so when we implement the API, users can switch to the new implementation when they are ready. So rotation wouldn't break immediately and we can keep it around for some time, so that  users can adapt their infra structure in time.

**Comment by MikeFear** [comment]
I support this proposal because I also ran into issues with the automatic key rotation and the proposal is in line how Auth0 handles it (I used Auth0 in previous projects). I'm using NestJS (Node.js) in the backend with Passport handling the authentication and there is the jwks-rsa library which adds support for JWKS and is widely used (over 3 mio. weekly downloads). The library is written and maintained by Auth0 and as they only use manual rotation it seems they don't handle the case of missing keys correctly: https://github.com/auth0/node-jwks-rsa/blob/master/src/JwksClient.js#L79

If the key doesn't exist it just throws an error that the key wasn't found and doesn't invalidate the cache which is enabled by default and caches all keys (probably even an empty key set) for 10 minutes by default. This causes the backend to regularly reject valid JWT tokens because the signing key was rotated but the new public key wasn't fetched yet. Took me quite some time to nail down the issue and understand what is actually happening :)

**Reply by fforootd** [reply]
Thanks for this input!

It is true that while rotation is a good thing it is not well supported in the clients 😓 
This was the reason to bring up this discussion and my feeling is we should switch to a manual key rotation with an api where people could automate the rotation if need be.

**Comment by jeffmccune** [comment]
I support this proposal.  There is prior art to support the decision as well, for example kubernetes itself does not expire signing keys by default.

For context, came here because of the issues noted integrating with istio `RequestAuthentication` as well.

**Comment by fforootd** [comment]
Since I got more and more requests in the community for this we should discuss if we are committed to the proposal and then a timeline.

My 2 cents are we should switch to a static key config with an api that can be automated since it resolves many of the pain points for the less intelligent clients 😁 

What do you think @muhlemmer @hifabienne (I also got a lot of questions for an ETA 😓 )

**Reply by hifabienne** [reply]
Yes I think we should implement this.
Will have to look with @muhlemmer what it means and make some estimations.

**Reply by muhlemmer** [reply]
Create an userstory: https://github.com/zitadel/zitadel/issues/7809

We might need to split it up. Also involves resources.

**Comment by ignatz** [comment]
Just my 2c, I don't mind the auto rotation, i think it's pretty elegant actually. I can certainly see how this would be a nuisance, if one relied on a 3rd party library that doesn't update its key cache. 

Anyway, to me the only issue is the lazy nature of minting, which makes periodic key refreshes inherently racy. If there simply was an eager, up-to-date sliding window of:
 * all keys that are still valid
 * **and** the next key to become valid

One could just periodically update the key cache race-free and be a happy camper.

I'm bring this up because:
 * it might be much simpler to implement, i.e. it doesn't require to upfront a new dashboard, build a more configurable key management flow, ...
 * might address the majority of issues
 * and I get to keep spinning :)

**Reply by danielloader** [reply]
Having read the thread and about to implement zitadel over keycloak in conjunction with istio for JWT authentication it's been mildly concerning.

The proposal to keep rotating the keys but with an overlap so tokens already issued have a natural way to expire and still be useful sounds like the "quick win" to minimise user issues with this deployment topology.

I'm not sure if istio/envoy will go out of their way to improve handling of the issue their end but much like kubernetes and other projects I was using keycloak with multi year tokens (with the infrequent but possible chance to revoke them on demand) and not thinking about it. 

Doubly so because the certificates were cycled on a schedule that suited us in a business context.

**Reply by danielloader** [reply]
Azure has an interesting documentation section on how they handle key rollover for similar reasons - https://learn.microsoft.com/en-us/entra/identity-platform/signing-key-rollover

**Comment by muhlemmer** [comment]
For subscribers to this topic, we released web key management and usage in v2.61.0. There is a usage guide here: https://zitadel.com/docs/guides/integrate/login/oidc/webkeys

Currently the solution is in alpha state and management is by API only. We would really like people to try it out and get some feedback at this stage. If something doesn't work, please open a bug report. Improvement discussions can go into a new GH discussion.

If the solution provides to be sufficiently stable we will promote the API to beta or GA directly and continue the implementation in the console.

Thanks in advance!

**Reply by fcenedes** [reply]
nice work on this, really appreciate the approach taken
